src/lib/controllers/ui-controller.js

```diff
@@ -45,11 +55,15 @@ export default class UIController {
    *                            popupComponent: Vue single file component of a panel element.
    *                              Allows to provide an alternative popup layout
    */
-  constructor (state, options, resourceOptions, uiOptions, manifest = {}, template = {}) {
-    this.state = state
-    this.options = options
-    this.resourceOptions = resourceOptions
-    this.uiOptions = uiOptions
+  constructor (storageAdapter, manifest = {}, template = {}) {
+    this.state = new TabScript()
```

Comment:
I think calling application has to identify the implementation of the UIStateAPI to be used.  TabScript makes sense for the webextension but not for the embedded library,for which state is local to the page. the PWA will also have its own State handling.  Implementations could live in components or with the application, as long as we can pick one upon construction.

Reply:
Agree. We don't have it implemented yet, but this can be useful for the future.

Reply:
Embedded library does have it's own state object right now, which does very little.  So we do need this to be able to be specified someway in the constructor now. It's okay with me if it's a provisional solution to start.

Reply:
I'm really afraid to touch any state objects right now because this can become a story on its own. I do believe that state handling needs some modernization but I really hope to postpone this for later, if possible.
If we make state a first argument of the constructor, would it solve the issue? Though I'm wondering if embedded lib's state is different then we probably need to have a different logic within a UI controller to handle webextension's state vs. embedded lib's state.

Reply:
I think the state object of the embedded library is largely non-functional right now. It mainly served to have an object that responded to methods called in the UIController so that it wouldn't crash :-)